lib/api/CDataFrameAnalyzer.cc

```diff
@@ -6,15 +6,133 @@
+}
+
+const std::string CONTROL_MESSAGE_FIELD_NAME{"."};
+// Control message types:
+const char RUN_ANALYSIS_CONTROL_MESSAGE_FIELD_VALUE{'r'};
+}
+
+// TODO memory calculations plus choose data frame storage strategy.
+CDataFrameAnalyzer::CDataFrameAnalyzer(CDataFrameAnalysisSpecification analysisSpecification)
+    : m_AnalysisSpecification{std::move(analysisSpecification)},
+      m_DataFrame{core::makeMainStorageDataFrame(m_AnalysisSpecification.cols())} {
+}
+
+bool CDataFrameAnalyzer::usingControlMessages() const {
+    return m_ControlFieldValue >= 0;
+}
 
 bool CDataFrameAnalyzer::handleRecord(const TStrVec& fieldNames, const TStrVec& fieldValues) {
-    LOG_INFO(<< core::CContainerPrinter::print(fieldNames));
-    LOG_INFO(<< core::CContainerPrinter::print(fieldValues));
+
+    // Control messages are signified by a dot in the field name. This supports:
+    //   - using the last field for a control message,
+    //   - missing.
+    //
+    // Note if the the control message field name is missing the analysis must
+    // be triggered to run by calling run explicitly.
+
+    bool result{true};
+    if (this->readyToReceiveControlMessages() == false) {
+        result = this->prepareToReceiveControlMessages(fieldNames);
+    }
+
+    if (this->isControlMessage(fieldValues)) {
+        return this->handleControlMessage(fieldValues);
+    }
+
+    this->addRowToDataFrame(fieldValues);
+
+    return result;
+}
+
+void CDataFrameAnalyzer::run() {
+    // TODO
+}
+
+bool CDataFrameAnalyzer::readyToReceiveControlMessages() const {
+    return m_ControlFieldValue != CONTROL_FIELD_UNSET;
+}
+
+bool CDataFrameAnalyzer::prepareToReceiveControlMessages(const TStrVec& fieldNames) {
+    if (fieldNames.empty() || fieldNames.back() != CONTROL_MESSAGE_FIELD_NAME) {
+        m_ControlFieldValue = CONTROL_FIELD_MISSING;
+        m_BeginDataFieldValues = 0;
+        m_EndDataFieldValues = static_cast<std::ptrdiff_t>(fieldNames.size());
+    } else {
+        m_ControlFieldValue = static_cast<std::ptrdiff_t>(fieldNames.size() - 1);
+        m_BeginDataFieldValues = 0;
+        m_EndDataFieldValues = m_ControlFieldValue;
+        auto pos = std::find(fieldNames.begin(), fieldNames.end(), CONTROL_MESSAGE_FIELD_NAME);
+        if (pos != fieldNames.end()) {
+            LOG_ERROR(<< "Unexpected possible control field: ignoring");
+            return false;
+        }
+    }
     return true;
 }
+
+bool CDataFrameAnalyzer::isControlMessage(const TStrVec& fieldValues) const {
+    return m_ControlFieldValue != CONTROL_FIELD_MISSING &&
+           fieldValues[m_ControlFieldValue].size() > 0;
+}
+
+bool CDataFrameAnalyzer::handleControlMessage(const TStrVec& fieldValues) {
+    bool unrecognised{fieldValues.size() > 1};
+    if (unrecognised == false) {
+        switch (fieldValues[m_ControlFieldValue][0]) {
+        case RUN_ANALYSIS_CONTROL_MESSAGE_FIELD_VALUE:
+            this->run();
+            break;
+        default:
+            unrecognised = true;
```

Comment:
It should also recognise a value of `.` and silently ignore the row.  (This allows for resending the field names.)

And it should also recognise a value of ` ` (space) and silently ignore the row.  This allows for flushing the input buffers.

To allow for code reuse on the Java side the logic for these should match https://github.com/elastic/ml-cpp/blob/master/lib/api/CAnomalyJob.cc#L314-L323

Reply:
Good point on flush.

Do we really want to allow for resend of field names via a stop? I can't think of a reason for needing to do this.

Reply:
If we ever keep the C++ process alive for further analysis then it means the code on the Java side that prepares the input doesn't need to know whether it's sending the very first batch to the process or a subsequent batch.  I agree it's not a requirement to add it now though - it can be added later if needed.

Reply:
I think in the context of sending data for multiple analyses the model will be a little different than for autodetect. What will likely happen will be that we'd send all the data, but potentially send subsequent requests to perform specific analyses. I think I'll leave this part for now, as there is no problem introducing later if we change our minds. I addressed the flush issue in [3291b12](https://github.com/elastic/ml-cpp/pull/323/commits/3291b12a9127e463bb43cd48d11c71066e738e41).

Reply:
I should add, there doesn't seem to be a large advantage in keeping a process running for analyses on different data frames to me at present, but we can think about pros and cons offline.